### DISCUSSION TITLE
Automatically expand volumes as needed

### DISCUSSION THREAD

**Discussion by vbezhenar** [cloudnative-pg/cloudnative-pg#494] [Ideas]
This project is awesome but I think that one feature would be very useful. According to the documentation, user is supposed to edit Cluster resource to expand volume as the database grows. I think that operator should monitor persistent volume usage and expand it as needed automatically (of course using some configured bounds, increments, etc). It'd allow to spend less money on storage (because you can start with small volume and use small increments) and spend less time monitoring this aspect of the system.

**Comment by sxd** [comment]
Hi @vbezhenar 

This sounds like a good feature to have, but can you develop more the idea and use case of these feature? Also, how you propose to this? Wouldn't this be part of some task that manage storage for every element inside the Kubernetes cluster instead of the job of one operator?

In the meantime I'm converting this into a discussion since it's more appropriate to discuss first and then create issues to work on.

Regards,

**Reply by vbezhenar** [reply]
Well, issue is that persistent volume have limited size and database usually grows with time. So someone should periodically check out volume usage and expand volume. It means that if that person forgot to do that, database will run out of disk space and stop working. Of course there should be metrics and alerts set up, but human errors happen. And it's mechanical work.

Another alternative is just allocate very big volume. But it means that someone will have to pay for unused space.

Postgres operator could monitor PV usage and expand it when it's about to become full. I expect it to run periodically with configurable schedule. When disk usage is over configurable threshold (e.g. 80%), operator should extend persistent volume size using configurable size increments (either relative like +5% or absolute like +1GiB) until this threshold is satisfied. And some absolute maximum should be specified to avoid situation where buggy application causes database grow infinitely.

There's still risk of database running out of disk space if it grows faster than job schedule allows. I don't know if it's possible to detect this event for operator and act outside of schedule. I guess it's nice to have that feature if it's possible.

I have no idea if universal solution exist for all Kubernetes cluster PVs. I guess that Postgres Operator should not touch other PVs that it does not manage.

**Reply by dhilipkumars** [reply]
I think that first we need the ability to [detect the current disk usage](https://github.com/cloudnative-pg/cloudnative-pg/discussions/495) before designing a volume auto-scaler feature.    CNP is agnostic to underlying disk implementation, each disks expansion work differently, some disk-expansion might require restart of pods etc (Azure had that limitation recently) which might result in a momentary disruption to the database.
